lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -0,0 +1,744 @@
+#include "llvm/ADT/MapVector.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "infer-widths"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Constraint Expressions
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct Expr;
+} // namespace
+
+/// Allow rvalue refs to `Expr` and subclasses to be printed to streams.
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const T &e) {
+  e.print(os);
+  return os;
+}
+
+/// Allow pointers to `Expr` and subclasses to be printed to streams.
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const T *e) {
+  return os << *e;
+}
+
+// Allow expression subclasses to be hashed.
+namespace mlir {
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::hash_code hash_value(const T &x) {
+  return x.hash_value();
+}
+} // namespace mlir
+
+namespace {
+
+/// An expression on the right-hand side of a constraint.
+struct Expr {
+  enum Kind {
+    KindNil,
+    KindVar,
+    KindKnown,
+    KindAdd,
+    KindMax,
+    KindMin,
+  };
+  Kind kind;
+  llvm::Optional<int32_t> solution = {};
+
+  /// Print a human-readable representation of this expr.
+  void print(llvm::raw_ostream &os) const;
+
+protected:
+  Expr(Kind kind) : kind(kind) {}
+  llvm::hash_code hash_value() const { return llvm::hash_value(kind); }
+};
+
+/// Helper class to CRTP-derive common functions.
+template <class DerivedT, Expr::Kind DerivedKind>
+struct ExprBase : public Expr {
+  ExprBase() : Expr(DerivedKind) {}
+  static bool classof(const Expr *e) { return e->kind == DerivedKind; }
+  bool operator==(const Expr &other) const {
+    if (auto otherSame = dyn_cast<DerivedT>(other))
+      return *this == otherSame;
+    return false;
+  }
+};
+
+/// The absence of an expression. Useful to represent types for which no solver
+/// variables are generated.
+struct NilExpr : public ExprBase<NilExpr, Expr::KindNil> {
+  void print(llvm::raw_ostream &os) const { os << "nil"; }
+  bool operator==(const NilExpr &other) const { return true; }
+};
+
+/// A free variable.
+struct VarExpr : public ExprBase<VarExpr, Expr::KindVar> {
+  VarExpr(unsigned index) : ExprBase(), index(index) {}
+  void print(llvm::raw_ostream &os) const { os << "var" << index; }
+  bool operator==(const VarExpr &other) const { return index == other.index; }
+  llvm::hash_code hash_value() const {
+    return Expr::hash_value() ^ llvm::hash_value(index);
+  }
+
+  /// A unique index for this variable.
+  unsigned index = 0;
```

Comment:
Could you turn off interning for VarExprs and use pointer identity?  It seems like a waste to put an index here, but I'm not really sure.

Reply:
Yeah that would make a lot of sense. I wasn't sure at the beginning if we would end up identifying variables through some other means and thus needed a way to intern them. But spitting out the same `VarExpr` multiple times is probably something to be handled in `InferenceMapping`, not the solver.